ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,329 @@
+                e_p.iter()
+                    .enumerate()
+                    .filter(|&j| j.0 != b.0)
+                    .for_each(|j| {
+                        *b.1 *= *point - *j.1;
+                    });
+            });
+        });
+}
+
+/// precomputed lagrange evaluation
+/// evaluates polynomial using base polynomials
+/// the base polynomials are tied to specific evaluation points
+/// multiplies `result` with input `result`, i.e. use input `F::ONE`
+pub fn lagrange_evaluation_precomputed<F>(y: &[F], base: &Vec<Vec<F>>, result: &mut [F])
+where
+    F: Field,
+{
+    debug_assert_eq!(base.len(), result.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base {
+            debug_assert_eq!(y.len(), base.len());
+        }
+    }
+    result
+        .iter_mut()
+        .zip(base.iter())
+        .for_each(|(result, base)| {
+            *result *= base
+                .iter()
+                .zip(y.iter())
+                .fold(F::ZERO, |acc, (base, y)| acc + *base * *y);
+        });
+}
+
+/// lagrange evaluation
+/// given points `(x_0,y_0),...,(x_n,y_n)` on a polynomial `p` and points `p_0,...,p_m`
+/// the function computes `p` evaluated on points `p_0,...,p_m`
+///
+/// It follows the lagrange method to compute the points `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// further, rather than outputting `p(p_k)`, we set `result_k=result_k*p(p_k)`
+/// which fits better to how we use `p(p_k)`
+///
+/// # Panics
+/// Panics when there are identical evaluation points, i.e. division by zero
+pub fn lagrange_evaluation<F>(x: &[F], y: &[F], p: &[F], result: &mut [F])
+where
+    F: Field + Invert,
+{
+    debug_assert_eq!(x.len(), y.len());
+    debug_assert_eq!(p.len(), result.len());
+    #[cfg(debug_assertions)]
+    {
+        for i in 0..x.len() {
+            for j in 0..x.len() {
+                assert_eq!((i, j, x[i] - x[j] != F::ZERO || j == i), (i, j, true));
+            }
+        }
+    }
+
+    // compute denominators:
+    let mut denominators = vec![F::ONE; x.len()];
+    for i in 0..x.len() {
+        if i > 0 {
+            for j in 0..i {
+                denominators[i] *= x[i] - x[j];
+            }
+        }
+        if i + 1 < x.len() {
+            for j in i + 1..x.len() {
+                denominators[i] *= x[i] - x[j];
+            }
+        }
+        denominators[i] = denominators[i].invert();
+    }
+
+    for k in 0..p.len() {
+        // evaluate polynomial on point p_k, i.e. compute `p(p_k)` use Lagrange formula
+        let mut sum = F::ZERO;
+        for i in 0..x.len() {
+            let mut basis_polynomial = denominators[i] * y[i];
+            if i > 0 {
+                for j in x.iter().take(i) {
+                    basis_polynomial *= p[k] - *j;
+                }
+            }
+            if i + 1 < x.len() {
+                for j in x.iter().skip(i + 1) {
+                    basis_polynomial *= p[k] - *j;
+                }
+            }
```

Comment:
Again, seems like it would be simpler to write:
```
for i in 0..x.len() {
  for j in 0..len() {
    if i != j {
      basis_polynomial *= p[k] - *j
    }
  }
}
```